### PR TITLE
Cuonglm/fix unhandle errors

### PR DESCRIPTION
**Pull request by Gnouc**

### CODE REVIEW

**Comment by Gnouc** [comment]
cc @thaidn

**Comment by thaidn** [comment]
@chuckx

File: go/subtle/util.go [hunk 1]
```diff
@@ -85,11 +88,18 @@ func GetCurve(curve string) elliptic.Curve {
 }
 
 // ComputeHash calculates a hash of the given data using the given hash function.
-func ComputeHash(hashFunc func() hash.Hash, data []byte) []byte {
+func ComputeHash(hashFunc func() hash.Hash, data []byte) (hashed []byte, err error) {
+	if hashFunc == nil {
```
Comment (thaidn) [inline go/subtle/util.go]: A nil function would fail anyway, I don't think this check is necessary.

File: go/subtle/util.go [hunk 2]
```diff
@@ -85,11 +88,18 @@ func GetCurve(curve string) elliptic.Curve {
...
+		err = errNilHashFunc
+		return
+	}
 	h := hashFunc()
-	h.Write(data)
-	ret := h.Sum(nil)
-	return ret
+	_, err = h.Write(data)
+	if err != nil {
+		return
```
Comment (thaidn) [inline go/subtle/util.go]: why don't you return the error here?

Comment (thaidn) [reply go/subtle/util.go]: Ah it seems that you already name err as the returned error.

I'm not sure why this style is better, but it's inconsistent with the rest of our library, so I'd prefer not doing it.

File: go/subtle/util_test.go [hunk 3]
```diff
@@ -41,25 +42,29 @@ func TestConvertCurveName(t *testing.T) {
...
 func TestComputeHash(t *testing.T) {
 	data := []byte("Hello")
-	// SHA1
-	expectedMac := "f7ff9e8b7bb2e09b70935a5d785e0cc5d9d0abf0"
-	hashFunc := subtle.GetHashFunc("SHA1")
-	if hashFunc == nil || hex.EncodeToString(subtle.ComputeHash(hashFunc, data)) != expectedMac {
-		t.Errorf("invalid hash function for SHA1")
+	var tests = []struct {
+		hashFunc    func() hash.Hash
+		expectedMac string
```
Comment (thaidn) [inline go/subtle/util_test.go]: there's a typo in the original code. this should be expectedHash

File: go/subtle/util.go (hunk 1 above)
Comment (Gnouc) [reply go/subtle/util.go]: We have to check nil, other wise calling `hashFunc()` below if `hashFunc` is nil will panic

File: go/subtle/util.go (hunk 2 above)
Comment (Gnouc) [reply go/subtle/util.go]: Named return parameters often use in sort function, for self document https://golang.org/doc/effective_go.html#named-results

Anyway, it does not matter in this case.

**Comment by Gnouc** [comment]
@thaidn updated

**Comment by Gnouc** [comment]
@thaidn updated

**Pull request merged**